pkg/apis/shared/namevalue.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2023 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package shared
+
+// +kubebuilder:object:generate=true
+
+import (
+	"strconv"
+
+	apiextensionsV1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+)
+
+// NameValuePair is a name/value pair.
+type NameValuePair struct {
+	// Name of the property.
+	Name string `json:"name"`
+	// Value of the property, in plaintext.
+	//+optional
+	Value DynamicValue `json:"value,omitempty"`
+	// SecretKeyRef is the reference of a value in a secret store component.
+	//+optional
+	SecretKeyRef SecretKeyRef `json:"secretKeyRef,omitempty"`
+	// EnvRef is the name of an environmental variable to read the value from.
+	//+optional
+	EnvRef string `json:"envRef,omitempty"`
```

Comment:
These 3 should be pointers, which are used to determine which has been selected. We should have validation the 1 and only 1 is non nil.

Reply:
I don't agree with the use of pointers here because these are all string values (or structs containig string values), so we don't need to distinguish between empty strings and non-present values (we treat empty strings as non-defined because we do not allow empty values).

Also, in K8s mode, both `secretKeyRef` and `value` can be non-empty: when using the K8s secret store, the Operator fetches the value of the secret and sends it to the runtime as well as `secretKeyRef`, so the runtime knows that the value was already fetched from the Operator.

Reply:
An empty string isn't a valid `value` either right now. Components expect all strings to be non-empty, and it's a convention used commonly in dapr/dapr and components-contrib.

Although I hear you, I don't think this change is needed now, or at least not in this PR since I'm not changing the existing behavior here.

Reply:
:+1: 

FWIW, an empty string value is permissible in Kubernetes Secrets, so there is some inconsistencies in API there.